d2/src/main/java/com/linkedin/d2/balancer/D2ClientBuilder.java

```diff
@@ -146,6 +146,8 @@ public D2Client build()
                   _config._executorService,
                   _config.retry,
                   _config.retryLimit,
+                  _config.maxClientRequestRetryRatio,
```

Comment:
There's only one D2Client per war making calls to different downstream service endpoints. I think it makes more sense to track retry ratio individually for each downstream endpoint. The latest commit uses a map to track retry info for each extracted service name.

The ratio should be defined by the server in the TransportClientProperties, because service owners have the best knowledge of how much retry traffic is allowed. This is also reflected in the latest commit.